<a href="https://colab.research.google.com/github/kishon45229/Customer-Churn-Prediction-in-Telecom-Industry/blob/main/Data_Mining.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Customer Churn Prediction in Telecom Industry





In the "Customer Churn Prediction in the Telecom Industry" project, I was responsible for the data mining aspect, which is crucial for extracting actionable insights from vast amounts of customer data. My role involved collecting, cleaning, and analyzing large datasets to identify patterns and trends that indicate potential customer churn. By applying various data mining techniques, such as clustering, classification, and association rule mining, I was able to uncover key factors contributing to customer attrition. This process enabled the development of predictive models that help telecom companies proactively address customer needs, ultimately reducing churn rates and improving customer retention strategies.

ITBIN-2110-0013

H.M.F.Asma

# **Data Mining**

# Train a model

### Import Libraries

In [67]:
import pandas as pd
from sklearn.preprocessing import LabelEncoder
from sklearn.model_selection import train_test_split
from sklearn.ensemble import RandomForestClassifier
from sklearn.metrics import classification_report, accuracy_score
from sklearn.model_selection import train_test_split
from sklearn.metrics import accuracy_score
from sklearn.datasets import load_iris
from sklearn.ensemble import RandomForestClassifier
from sklearn.svm import SVC
from sklearn.neighbors import KNeighborsClassifier
from sklearn.tree import DecisionTreeClassifier
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import StandardScaler, OneHotEncoder, LabelEncoder
from sklearn.impute import SimpleImputer
from sklearn.compose import ColumnTransformer
from sklearn.pipeline import Pipeline
from sklearn.feature_selection import SelectKBest, f_classif
from sklearn.ensemble import RandomForestClassifier
from sklearn.metrics import accuracy_score
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import StandardScaler, LabelEncoder
from sklearn.preprocessing import OneHotEncoder, StandardScaler
from sklearn.metrics import accuracy_score, classification_report
import matplotlib.pyplot as plt
import numpy as np


## Load the File

In [15]:
new_file_path = '/content/Section 3 finished dataset.csv'
new_data = pd.read_csv(new_file_path)

# Random Forest

### Splitting the data into features and target variable

In [6]:

X = new_data.drop('Churn', axis=1)
y = new_data['Churn']
# in here im going to Splitting the data into features and target variable


### Splitting the data into training and testing sets

In [7]:

X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.3, random_state=42)
#split the data into training (70%) and testing (30%) sets using train_test_split.
#The random_state parameter ensures reproducibility.

Create an instance of RandomForestClassifier with a fixed random_state for reproducibility.
it the classifier on the training data (X_train and y_train).

In [8]:
# Initializing and training the RandomForestClassifier
clf = RandomForestClassifier(random_state=42)
clf.fit(X_train, y_train)


RandomForestClassifier(random_state=42)

The model will be trained on the data to create an ensemble of decision trees for classification tasks, leading to more accurate and stable predictions compared to a single decision tree.

## Decision Tree model.

trained the Decision Tree model.

In [58]:
model = DecisionTreeClassifier(random_state=42)
model.fit(X_train, y_train)

y_pred = model.predict(X_test)

evaluate the Decision Tree model's performance

In [59]:
accuracy = accuracy_score(y_test, y_pred)
print(f'Accuracy: {accuracy:.2f}')

print(classification_report(y_test, y_pred))

Accuracy: 1.00
              precision    recall  f1-score   support

           0       1.00      1.00      1.00        10
           1       1.00      1.00      1.00         9
           2       1.00      1.00      1.00        11

    accuracy                           1.00        30
   macro avg       1.00      1.00      1.00        30
weighted avg       1.00      1.00      1.00        30



Accuracy is the proportion of correct predictions (both true positives and true negatives) out of the total number of predictions. It is a measure of the overall effectiveness of the model.

I here it shows the accuracy 1.00, the model correctly predicted the class labels for all 30 instances in the dataset. by using decision tree model

Use the trained classifier to make predictions on the test set (X_test).

In [9]:
# Making predictions on the test set
y_pred = clf.predict(X_test)

# Evaluating the model
accuracy = accuracy_score(y_test, y_pred)
report = classification_report(y_test, y_pred)

print(f"Accuracy: {accuracy}")
print("Classification Report:\n", report)


Accuracy: 0.7867298578199052
Classification Report:
               precision    recall  f1-score   support

           0       0.83      0.90      0.86      1549
           1       0.63      0.48      0.54       561

    accuracy                           0.79      2110
   macro avg       0.73      0.69      0.70      2110
weighted avg       0.77      0.79      0.78      2110



 Check for missing values

In [10]:
missing_values = new_data.isnull().sum()
print("Missing values in each column:\n", missing_values)

Missing values in each column:
 customerID            0
gender                0
SeniorCitizen         0
Partner               0
Dependents            0
tenure                0
PhoneService          0
MultipleLines         0
InternetService       0
OnlineSecurity        0
OnlineBackup          0
DeviceProtection      0
TechSupport           0
StreamingTV           0
StreamingMovies       0
Contract              0
PaperlessBilling      0
PaymentMethod         0
MonthlyCharges        0
TotalCharges          0
Churn                 0
tenure_bin            0
MonthlyCharges_bin    0
Cluster               0
PCA1                  0
PCA2                  0
dtype: int64


In [22]:
print("First 5 rows of X_train:\n", X_train[:5])
print("First 5 rows of X_test:\n", X_test[:5])
print("First 5 values of y_train:\n", y_train[:5])
print("First 5 values of y_test:\n", y_test[:5])

First 5 rows of X_train:
       SeniorCitizen  tenure  MonthlyCharges  TotalCharges  Cluster      PCA1  \
6021              0      43           49.05       2076.20        0 -0.422284   
3404              0       3           53.40        188.70        2 -1.277010   
5474              0      55           77.75       4458.15        1  1.203379   
5515              0      45           54.65       2553.70        0 -0.147057   
6328              0      55          100.90       5448.60        1  1.900721   

          PCA2  gender_Male  Partner_Yes  Dependents_Yes  ...  \
6021 -1.429241        False        False           False  ...   
3404  1.310409         True        False           False  ...   
5474 -0.661804        False         True           False  ...   
5515 -1.441230         True         True            True  ...   
6328 -0.455413        False         True            True  ...   

      tenure_bin_13-24  tenure_bin_25-36  tenure_bin_37-48  tenure_bin_49-60  \
6021             False

This is the target variable corresponding to the training dataset, indicating whether the customer churned (1) or not (0).

The columns are identical to those in X_train, containing similar customer attributes, but the data belongs to the test set.

## **Model Evaluation**

In [25]:
from sklearn.metrics import accuracy_score, precision_score, recall_score, f1_score, confusion_matrix, classification_report

# Predict on the test set
y_pred = model.predict(X_test)

# Evaluate the model
accuracy = accuracy_score(y_test, y_pred)
precision = precision_score(y_test, y_pred)
recall = recall_score(y_test, y_pred)
f1 = f1_score(y_test, y_pred)
conf_matrix = confusion_matrix(y_test, y_pred)
class_report = classification_report(y_test, y_pred)

/usr/local/lib/python3.10/dist-packages/sklearn/base.py:465: UserWarning: X does not have valid feature names, but RandomForestClassifier was fitted with feature names
  warnings.warn(


Print evaluation metrics

In [26]:
print(f"Accuracy: {accuracy:.4f}")
print(f"Precision: {precision:.4f}")
print(f"Recall: {recall:.4f}")
print(f"F1-score: {f1:.4f}")

Accuracy: 0.6766
Precision: 0.4389
Recall: 0.7781
F1-score: 0.5612


Accuracy (67.66%): The model correctly predicted about two-thirds of the instances.

Precision (43.89%): The model's positive predictions were less than half correct, indicating a higher rate of false positives.

Recall (77.81%): The model identified more than three-quarters of the actual positive instances but missed some, indicating some false negatives.

F1-Score (56.12%): The F1-score shows a moderate balance between precision and recall, suggesting the model has room for improvement in balancing these metrics.



Print confusion matrix

In [27]:
print("Confusion Matrix:")
print(conf_matrix)

Confusion Matrix:
[[661 372]
 [ 83 291]]


Print classification report

In [28]:
print("Classification Report:")
print(class_report)

Classification Report:
              precision    recall  f1-score   support

           0       0.89      0.64      0.74      1033
           1       0.44      0.78      0.56       374

    accuracy                           0.68      1407
   macro avg       0.66      0.71      0.65      1407
weighted avg       0.77      0.68      0.70      1407



# **Experiment with different parameter settings**

In [29]:
from sklearn.model_selection import train_test_split
from sklearn.ensemble import RandomForestClassifier
from sklearn.metrics import accuracy_score, classification_report
from sklearn.preprocessing import LabelEncoder

In [30]:
from sklearn.metrics import accuracy_score, precision_score, recall_score, f1_score, confusion_matrix, classification_report

Predict on the test set

In [31]:
y_pred = model.predict(X_test)

/usr/local/lib/python3.10/dist-packages/sklearn/base.py:465: UserWarning: X does not have valid feature names, but RandomForestClassifier was fitted with feature names
  warnings.warn(


Evaluate the model

In [32]:
accuracy = accuracy_score(y_test, y_pred)
precision = precision_score(y_test, y_pred)
recall = recall_score(y_test, y_pred)
f1 = f1_score(y_test, y_pred)
conf_matrix = confusion_matrix(y_test, y_pred)
class_report = classification_report(y_test, y_pred)

Print evaluation metrics

In [33]:
print(f"Accuracy: {accuracy:.4f}")
print(f"Precision: {precision:.4f}")
print(f"Recall: {recall:.4f}")
print(f"F1-score: {f1:.4f}")

Accuracy: 0.6766
Precision: 0.4389
Recall: 0.7781
F1-score: 0.5612


Print confusion matrix

In [34]:
print("Confusion Matrix:")
print(conf_matrix)

Confusion Matrix:
[[661 372]
 [ 83 291]]


The confusion matrix is a table used to evaluate the performance of a classification model. It summarizes the counts of true positives, false positives, true negatives, and false negatives

Print classification report

In [35]:
print("Classification Report:")
print(class_report)

Classification Report:
              precision    recall  f1-score   support

           0       0.89      0.64      0.74      1033
           1       0.44      0.78      0.56       374

    accuracy                           0.68      1407
   macro avg       0.66      0.71      0.65      1407
weighted avg       0.77      0.68      0.70      1407



Predictions: The model predicts the target variable on the test set.
Metrics: Accuracy, precision, recall, and F1-score are computed to evaluate model performance.
Confusion Matrix: Displays the number of true positives, true negatives, false positives, and false negatives.
Classification Report: Provides a detailed performance report, including precision, recall, and F1-score for each class.

## Experiment with multiple alternative methods

 Load dataset (for example, Iris dataset)

In [36]:
data = load_iris()
X = data.data
y = data.target

Split the data into training and testing sets

In [37]:
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

Initialize different classifiers

In [38]:
classifiers = {
    "Random Forest": RandomForestClassifier(),
    "Support Vector Machine": SVC(),
    "K-Nearest Neighbors": KNeighborsClassifier(),
    "Decision Tree": DecisionTreeClassifier()
}

Experiment with different classifiers

In [39]:
for name, clf in classifiers.items():
    clf.fit(X_train, y_train)  # Train the classifier
    y_pred = clf.predict(X_test)  # Make predictions
    accuracy = accuracy_score(y_test, y_pred)  # Evaluate accuracy
    print(f"{name} Accuracy: {accuracy:.2f}")

Random Forest Accuracy: 1.00
Support Vector Machine Accuracy: 1.00
K-Nearest Neighbors Accuracy: 1.00
Decision Tree Accuracy: 1.00


This code loads a dataset (in this case, the Iris dataset), splits it into training and testing sets, and then trains and evaluates different machine learning classifiers (Random Forest, SVM, KNN, and Decision Tree) to compare their performance.

# **Improve preprocessing and feature generation**

Split the data into training and testing sets

In [40]:
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

Define numerical and categorical features

In [41]:
df = pd.DataFrame({
    'Category': ['A', 'B', 'A', 'C', 'B'],
    'Value': [1.0, 2.5, 3.3, 4.4, 2.2]
})


Define Transformer

In [42]:
numeric_features = ['Value']
categorical_features = ['Category']

Preprocessing for numerical data

In [43]:
numeric_transformer = Pipeline(steps=[
    ('imputer', SimpleImputer(strategy='mean')),  # Handle missing values
    ('scaler', StandardScaler())  # Normalize the data
])


Categorical transformer pipeline: Imputation + OneHotEncoding

In [44]:
categorical_transformer = Pipeline(steps=[
    ('imputer', SimpleImputer(strategy='most_frequent')),  # Handle missing values
    ('onehot', OneHotEncoder())  # Encode categorical features
])


Preprocessing for categorical data

In [45]:
categorical_pipeline = Pipeline(steps=[
    ('imputer', SimpleImputer(strategy='most_frequent')),  # Handle missing values
    ('onehot', OneHotEncoder(handle_unknown='ignore'))  # One-hot encoding
])

Combine the transformers using ColumnTransformer

In [46]:
preprocessor = ColumnTransformer(
    transformers=[
        ('num', numeric_transformer, numeric_features),
        ('cat', categorical_transformer, categorical_features)
    ])

Apply the transformations

In [47]:
X = preprocessor.fit_transform(df)

print(X)

[[-1.48168659  1.          0.          0.        ]
 [-0.15875213  0.          1.          0.        ]
 [ 0.54681291  1.          0.          0.        ]
 [ 1.51696484  0.          0.          1.        ]
 [-0.42333903  0.          1.          0.        ]]


The first column appears to be standardized, as the values range around 0 and do not seem to follow a specific unit or scale typical for raw data

The second, third, and fourth columns represent one-hot encoded features. This technique is used to convert categorical data into a format that can be provided to ML algorithms to do a better in prediction

Feature generation using feature selection

In [49]:
feature_generation_pipeline = Pipeline(steps=[
    ('preprocessor', preprocessor),
    ('feature_selection', SelectKBest(score_func=f_classif, k=10))  # Select top 10 features
])


 Train a model with the improved preprocessing and feature generation

In [50]:
model_pipeline = Pipeline(steps=[
    ('feature_gen', feature_generation_pipeline),
    ('classifier', RandomForestClassifier(random_state=42))
])


Numerical Data: Missing values are handled using the median, and the features are scaled using StandardScaler.
Categorical Data: Missing values are handled using the most frequent value, and the features are encoded using one-hot encoding.


 Make predictions

In [54]:
# Predict on the test data
y_pred = model.predict(X_test)

# Evaluate the model
accuracy = accuracy_score(y_test, y_pred)
report = classification_report(y_test, y_pred)

print(f"Accuracy: {accuracy:.4f}")
print("Classification Report:")
print(report)


Accuracy: 1.0000
Classification Report:
              precision    recall  f1-score   support

           0       1.00      1.00      1.00        10
           1       1.00      1.00      1.00         9
           2       1.00      1.00      1.00        11

    accuracy                           1.00        30
   macro avg       1.00      1.00      1.00        30
weighted avg       1.00      1.00      1.00        30



Simple Dataset: The dataset might be simple or linearly separable, making it easier for the model to achieve perfect classification.

Evaluate the model

In [55]:
accuracy = accuracy_score(y_test, y_pred)
print(f"Model accuracy after preprocessing and feature generation: {accuracy:.2f}")

Model accuracy after preprocessing and feature generation: 1.00


# Increase amount or quality of training data

## Data Augmentation